Add mechanism for retrieving BlockEntity data

glorantq opened the pull request:
This commit adds a mechanism for retrieving block entity data. Block entity data is required to support for example text on signs, banner patterns, or mods such as Domum Ornamentum.

glorantq committed:
Add mechanism for retrieving BlockEntity data

This commit adds a mechanism for retrieving block entity data.
Block entity data is required to support for example text on signs,
banner patterns, or mods such as Domum Ornamentum.

TBlueF commented on BlueMapCore/src/main/java/de/bluecolored/bluemap/core/world/block/BlockEntity.java:
```diff
@@ -0,0 +1,142 @@
+public class BlockEntity {
+    private static final List<String> COMMON_KEYS = List.of("id", "x", "y", "z", "keepPacked");
+
+    private final String id;
```
this can be a `Key` instead of a `String`

glorantq replied:
Not needed with the class-hierarchy approach

TBlueF commented on BlueMapCore/src/main/java/de/bluecolored/bluemap/core/world/block/BlockEntity.java:
```diff
@@ -0,0 +1,142 @@
+import java.util.stream.Collectors;
+
+public class BlockEntity {
+    private static final List<String> COMMON_KEYS = List.of("id", "x", "y", "z", "keepPacked");
```
```suggestion
    private static final Set<String> COMMON_KEYS = Set.of("id", "x", "y", "z", "keepPacked");
```

glorantq replied:
Not needed with the class-hierarchy approach

TBlueF commented on BlueMapCore/src/main/java/de/bluecolored/bluemap/core/world/block/Block.java:
```diff
@@ -147,6 +149,10 @@ public int getBlockLightLevel() {
         return getLightData().getBlockLight();
     }
 
+    public Optional<BlockEntity> getBlockEntity() {
```
Using an `Optional` is definitely the better practice, but i like to just return a nullable `BlockEntity` + `@Nullable` annotation here, to micro-optimize the extra wrapper object away ^^'

glorantq replied:
Done

TBlueF commented on BlueMapCore/src/main/java/de/bluecolored/bluemap/core/world/block/BlockEntity.java:
```diff
@@ -0,0 +1,142 @@
+
+    public String getString(String key, String def) {
+        return getString(data, key, def);
+    }
```
This doesn't feel right..
A lot of those methods could be `static`, which then begs the question why we need them in the first place, they look like a lenient but expensive way of casting any "object" into the value-type we need.

I understand the issue that you try to solve here: each tile-entity could have a different data-structure and thus can't be represented by one object-type with a fixed set of properties.

My approach for this would be to create a tree of fixed types anyways (e.g. `SignBlockEntity extends BlockEntity`) with well defined properties, and then write a [custom `TypeDeserializer<BlockEntity>`](https://github.com/BlueMap-Minecraft/BlueMap/tree/master/BlueMapCore/src/main/java/de/bluecolored/bluemap/core/world/mca/data) which first reads the whole data-tree (i don't think this can be avoided) but then (based on the `id`) creates the correct sub-type of `BlockEntity`.

Also i think instead of trying to convert between value-types (e.g. `Integer.parseInt(String.valueOf(obj))`) we should just cast based on the type that we expect (`(int) obj`) and deal with the cast-exception and treat it as invalid data-format.

glorantq replied:
Done

TBlueF commented on BlueMapCore/src/main/java/de/bluecolored/bluemap/core/world/mca/chunk/Chunk_1_13.java:
```diff
@@ -195,6 +204,13 @@ public int getOceanFloorY(int x, int z) {
+        return blockEntities.stream()
+                .filter(it -> it.getX() == x && it.getY() == y && it.getZ() == z)
+                .findFirst();
+    }
```
(return `@Nullable BlockEntity`, see comment above)

And we should definitely try to avoid any type of object-creation here (e.g. streams) and consider using some kind of hash-lookup or array-index to not have to loop over **all** tile-entities.
People have been filling chunks with flower-pot's already and tried to render them with bluemap so i can well imagine someone having hundreds of signs as well ^^

Maybe just splitting them into the individual chunk-sections could be a good-enough approach already 🤔

glorantq replied:
Done, replaced with packing the coordinates into a long and using that as a key for a Map. The only way to make this faster would be to make a custom map implementation that eliminates the small impact of the boxing, or bring over one of the `Long2Object*` maps from fastutil.

TBlueF reviewed (changes requested):
Thanks for the PR!
Just a really quick and hasty review for now, since i don't have much time rn until tuesday :)

TBlueF commented on BlueMapCore/src/main/java/de/bluecolored/bluemap/core/world/mca/chunk/Chunk_1_13.java:
```diff
@@ -273,6 +289,7 @@ public static class Level {
         private HeightmapsData heightmaps = new HeightmapsData();
         private SectionData @Nullable [] sections = null;
         private int[] biomes = EMPTY_INT_ARRAY;
+        @NBTName("TileEntities") private List<Map<String, Object>> blockEntities = List.of(Map.of());
```
With the `TypeDeserializer<BlockEntity>` mentioned above, this can then be simplified into a `List<BlockEntity>` :)

glorantq replied:
Done

TBlueF commented on BlueMapCore/src/main/java/de/bluecolored/bluemap/core/world/block/entity/BlockEntity.java:
```diff
@@ -0,0 +1,111 @@
+@NBTDeserializer(BlockEntity.BlockEntityDeserializer.class)
+public class BlockEntity {
+    private static final BlueNBT BLUENBT = new BlueNBT();
+    private static final Map<String, Class<? extends BlockEntity>> ID_MAPPING = Map.of(
```
How about switching this for a `Map<String, Function<Map<String, Object>, ? extends BlockEntity>>` instead?
Then we can store the type's constructor in the map directly and don't need to use reflection below :)

glorantq replied:
Right, apart from the ungodly type we can just store a method reference to the constructor

TBlueF replied:
Can always write a small functional interface to make the type more readable :D

```java
Map<String, BlockEntityConstructor> ID_MAPPING = Map.of(
     "minecraft:sign", SignBlockEntity::new,
     "minecraft:skull", SkullBlockEntity::new,
     "minecraft:banner", BannerBlockEntity::new
);

@FunctionalInterface
private interface BlockEntityConstructor {
    BlockEntity construct(Map<String, Object> data);
}
```
super fancy :)

TBlueF commented on BlueMapCore/src/main/java/de/bluecolored/bluemap/core/world/mca/chunk/Chunk_1_13.java:
```diff
@@ -113,6 +119,10 @@ public Chunk_1_13(MCAWorld world, Data data) {
         }
+
+        this.blockEntities = level.blockEntities.stream().collect(Collectors.toMap(
+                it -> (long) it.getY() << 40 | (it.getX() & 0xF) << 4 | it.getZ() & 0xF, it -> it
```
I really like this solution for quick access!
Can you explain why y is shifted by `40` tho? I would have expected `8` or maybe `32` 🤔

glorantq replied:
You are right, it should be 8. It was late, the 40 came from 8 (x and z) + 32 (y) because for some reason I thought there that the shift moves the most significant bit. Oops.

glorantq committed:
Fix the coordinate-packing for block entity-loading

This commit fixes the incorrect shifting of bits when
packing the chunk-local coordinates of a block entity
into a 64-bit long for lookups.

glorantq committed:
Change mapping type of BlockEntity lookups

This commit changes the type stored for BlockEntity
mappings from a class of the type associated with the
ID to a method reference to its constructor.

glorantq committed:
Tidy BlockEntity mappings

This commit introduces a small functional interface
to make the type less ungodly. Also silences the warning
about referencing subclasses in the superclass, it is
fine in this case, we're just storing a reference to
the constructor.

TBlueF commented:
The github-build fails, but i think it only needs a `./gradlew spotlessApply` to add the license headers ^^

glorantq committed:
Add missing license headers

The license headers were missing. Oops.

glorantq commented:
Yeah I realised I missed the headers, should be fine now. `spotlessCheck` is happy locally.
